frontend/modules/issues/issues.redux.ts

```diff
@@ -0,0 +1,182 @@
+	fetchIssues: ['teamspace', 'modelId', 'revision'],
+	fetchIssuesSuccess: ['issues'],
+	fetchIssue: ['teamspace', 'modelId', 'issueId'],
+	fetchIssueSuccess: ['issue'],
+	fetchIssueFailure: [],
+	setComponentState: ['componentState'],
+	saveIssue: ['teamspace', 'model', 'issueData', 'revision'],
+	updateIssue: ['teamspace', 'modelId', 'issueData'],
+	postComment: ['teamspace', 'modelId', 'issueData'],
+	removeComment: ['teamspace', 'modelId', 'issueData'],
+	saveIssueSuccess: ['issue'],
+	setNewIssue: [],
+	renderPins: [],
+	printIssues: ['teamspace', 'modelId'],
+	downloadIssues: ['teamspace', 'modelId'],
+	showDetails: ['teamspace', 'model', 'revision', 'issue'],
+	closeDetails: ['teamspace', 'model', 'revision'],
+	setActiveIssue: ['issue', 'revision'],
+	showNewPin: ['issue', 'pinData'],
+	togglePendingState: ['isPending'],
+	toggleDetailsPendingState: ['isPending'],
+	subscribeOnIssueChanges: ['teamspace', 'modelId'],
+	unsubscribeOnIssueChanges: ['teamspace', 'modelId'],
+	focusOnIssue: ['issue', 'revision'],
+	toggleIsImportingBcf: ['isImporting'],
+	toggleSubmodelsIssues: ['showSubmodelIssues'],
+	importBcf: ['teamspace', 'modelId', 'file', 'revision'],
+	exportBcf: ['teamspace', 'modelId'],
+	subscribeOnIssueCommentsChanges: ['teamspace', 'modelId', 'issueId'],
+	unsubscribeOnIssueCommentsChanges: ['teamspace', 'modelId', 'issueId'],
+	createCommentSuccess: ['comment', 'issueId'],
+	deleteCommentSuccess: ['commentGuid', 'issueId'],
+	updateCommentSuccess: ['comment', 'issueId'],
+	toggleSortOrder: ['sortOrder'],
+	updateNewIssue: ['newIssue'],
+	setFilters: ['filters'],
+	showCloseInfo: ['issueId'],
+	resetComponentState: []
+}, { prefix: 'ISSUES_' });
+
+export const INITIAL_STATE = {
+	issuesMap: {},
+	isPending: true,
+	componentState: {
+		activeIssue: null,
+		showDetails: false,
+		expandDetails: true,
+		newIssue: {},
+		newComment: {},
+		selectedFilters: [],
+		filteredRisks: [],
+		showPins: true,
+		fetchingDetailsIsPending: false,
+		isImportingBCF: false,
+		showSubmodelIssues: false,
+		sortOrder: 'desc',
+		failedToLoad: false
+	}
+};
+
+export const togglePendingState = (state = INITIAL_STATE, { isPending }) => ({ ...state, isPending });
+
+export const toggleDetailsPendingState = (state = INITIAL_STATE, { isPending }) => {
+	return setComponentState(state, { componentState: { fetchingDetailsIsPending: isPending } });
+};
+
+export const toggleIsImportingBcf = (state = INITIAL_STATE, { isImporting }) => {
+	return setComponentState(state, { componentState: { isImportingBCF: isImporting }});
+};
+
+export const fetchIssuesSuccess = (state = INITIAL_STATE, { issues = [] }) => {
+	const issuesMap = keyBy(issues, '_id');
+	return {
+		...state, issuesMap, componentState: { ...cloneDeep(INITIAL_STATE.componentState) }
+	};
+};
+
+export const fetchIssueSuccess = (state = INITIAL_STATE, { issue }) => {
+	const issuesMap = cloneDeep(state.issuesMap);
+	issuesMap[issue._id].comments = issue.comments;
```

Comment:
Deep cloning is being used everywhere in this pr to update state. I think this might not right and will affect perfomance.

from this article:
https://redux.js.org/faq/performance#do-i-have-to-deep-clone-my-state-in-a-reducer-isnt-copying-my-state-going-to-be-slow

![image](https://user-images.githubusercontent.com/42032433/55622043-17a6db00-5797-11e9-9719-d30e5bee2543.png)

How to update state the proper way:
https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns

Reply:
Good point!  👍

Reply:
Fixed